Iterations and filter tuning

lalithsuresh opened the pull request:
This includes several fixes to reduce allocations from inefficient iterations. Much of it involves iterator re-use using fastutils collections, or avoiding keySet() based iterations over maps.

Please review carefully to make sure I didn't mess up the logic!

mbudiu-vmw commented on platform/src/main/java/org/hillview/sketches/FreqKList.java:
```diff
@@ -48,18 +51,18 @@
      * (can be approximate depending on the context).
      */
-    public final HashMap<RowSnapshot, Integer> hMap;
+    public final Object2IntOpenHashMap<RowSnapshot> hMap;
```
This is an interesting case. In general it should not make a difference optimizing sketch results, such as this one, because they are by definition small. However, this one is special, since the "filter" method is also used for a very expensive computation, so it does make sense to optimize it. This class is both a sketch result and part another sketch computation.

lalithsuresh replied:
I run things end-to-end from the GUI, so yes, didn't see anything fishy.

mbudiu-vmw commented on platform/src/main/java/org/hillview/utils/DateParsing.java:
```diff
@@ -123,11 +124,11 @@ public DateParsing(String s) {
-        for (String regexp : DATE_FORMAT_REGEXPS.keySet()) {
-            if (s.toLowerCase().matches(regexp)) {
-                String format = DATE_FORMAT_REGEXPS.get(regexp);
+        for (Map.Entry<String, String> regexpEntry : DATE_FORMAT_REGEXPS.entrySet()) {
```
This loop has around 40 iterations, so this is overkill.

lalithsuresh replied:
Doesn't hurt. :)

mbudiu-vmw commented:
Maybe you can rebase before we merge; there are some conflicts.

lalithsuresh commented on platform/src/main/java/org/hillview/dataset/RemoteDataSet.java:
```diff
@@ -58,9 +62,13 @@ public RemoteDataSet(final HostAndPort serverEndpoint) {
     public RemoteDataSet(final HostAndPort serverEndpoint, final int remoteHandle) {
         this.serverEndpoint = serverEndpoint;
         this.remoteHandle = remoteHandle;
+        final ExecutorService executorService = Executors.newFixedThreadPool(5);
```
Will do.

lalithsuresh commented:
Rebased and added two more fixes:
- Reduced boxing overheads (106MB -> 150KB) for filters.
- Avoid a copying step used for FreqKlist.filter().

 Didn't fix the executor usage yet. I have some optimizations in there to try out but not in this PR.
